### PR TITLE
fix: bound orchestrated-v2 analysis cost (#470)

### LINKED ISSUE #470: fix(orchestrated-v2): sub-tasks loop on read_tool_result_artifact and exhaust budget without finalize_subtask
## Summary

In `orchestrated-v2`, sub-tasks frequently hit the `read_tool_result_artifact` re-fetch loop and exhaust their token budget without ever calling `finalize_subtask`. The strategist gets back `BUDGET_EXHAUSTED` with no summary, falls back to the first tool's output preview (often an empty `kd_read_toc`), learns nothing, and either dispatches another batch of sub-tasks or gives up. ~$0.30–0.50 of Opus tokens burn per failed sub-task.

## Reproducibility

Live-observed during analysis of ticket `cf1b96e8-8835-4192-9964-cc66af385675` (Evolution DB deadlocks scheduled probe), 2026-04-27.

Sub-task pattern split was reliable across all 5 orchestrator iterations watched:

| Pattern | Outcome |
|---|---|
| Source-code analysis (one `read_file`, then summarize) | **Finalizes** — st-11, st-14, st-19, st-22 |
| `read_tool_result_artifact` re-fetch loop on cached MCP results | **BUDGET_EXHAUSTED, no summary** — st-7, 8, 10, 12, 13, 15, 20, 21, 23 |

Approximate failure rate: ~60% of sub-tasks across the watched window. Cost burn for the watched window alone: rough estimate $5–8 across iterations 2–5.

## Mechanism (hypothesis)

The sub-task agent receives an MCP tool result that is large enough to be saved as an artifact (so the actual tool result is replaced with a pointer + head/tail preview). The agent then calls `read_tool_result_artifact` to "expand" the result. Because the read returns the same content the agent already has in context (just confirmed), the agent calls again on a different artifact — and again — and again. There's no negative signal that "you already saw this." It loops until budget is exhausted.

The system prompt likely doesn't push hard enough toward `finalize_subtask` once the agent has gathered "enough" context, AND the truncation pointer model encourages re-reading.

## Suggested investigations

1. **Per-sub-task system prompt** — does it tell the agent: "you have a finite budget, call `finalize_subtask` early with what you have"? If not, add explicit budget pressure.
2. **`read_tool_result_artifact` semantics** — should it short-circuit when the agent's last N tool calls already include the requested artifact? Or warn the agent in the response: "you've already read this — call finalize_subtask"?
3. **Iteration-cap vs token-cap interaction** — sub-tasks are exhausting on tokens, not iterations. Maybe lower the per-tool-result truncation threshold so artifacts get inlined more aggressively (avoiding the read tool altogether for small results).
4. **Auto-finalize on budget warning** — if the next iteration's expected token cost would push us over budget, force a finalize-or-summarize step instead of continuing.

## Cost impact

For a probe-sourced ticket like `cf1b96e8`, the analyzer was on track to spend ~$15–25 of Opus tokens with ~60% of that being unproductive sub-task budget burn before the operator manually killed the job.

## Files

- `services/ticket-analyzer/src/analysis/orchestrated-v2.ts` — sub-task dispatch + result handling
- `services/ticket-analyzer/src/analysis/v2-subtask-tools.ts` — `dispatch_subtasks`, `finalize_subtask` definitions
- `mcp-servers/platform/src/tools/read-tool-result-artifact.ts` — the tool agents are looping on
- `services/ticket-analyzer/src/analysis/shared.ts` — `SUB_TASK_ITERATION_CAP` / token-budget logic

## Origin

Discovered during live-watch of in-flight analysis on 2026-04-27. Job manually cancelled before completion.

### PR DESCRIPTION
**Pull request by siir**
## Summary

Caps worst-case orchestrated-v2 ticket-analyzer cost at the configured ticket budget (default 300k tokens ≈ $5). Resolves #470.

Adds 5 layered guardrails:
- **A** — budget-aware sub-task system prompt
- **B** — sub-task budget soft-nudge (60%) + hard-stop (85%)
- **C** — per-artifact re-read detector
- **D** — strategist batch-failure guard (cumulative + consecutive metrics)
- **E** — ticket-level total-token cap with continuation-summary directive on hard-stop (E.1)

All thresholds runtime-configurable via the new `orchestrated-v2-budget-config` AppSetting, surfaced in Settings → Analysis tab.

Spec: `docs/superpowers/specs/2026-04-28-orchestrated-v2-budget-bounds-design.md`
Plan: `docs/superpowers/plans/2026-04-28-orchestrated-v2-budget-bounds.md`

Deferred follow-ups: #475 (MCP tool pair), #476 (per-category overrides).

## Layers — what each one does

| Layer | Where | Trigger | Effect |
|---|---|---|---|
| A | `runSubTaskLoop` system prompt | n/a (static) | Replaces the old "call finalize_subtask LAST" instruction (which encouraged budget exhaustion) with budget-aware guidance: warn about thresholds, prefer grep over paging, partial findings preferable to budget burn |
| B (soft) | top of each sub-task iteration | 60% of any axis (tokens / iterations / calls) | Inject user-text warning into messages (fires once per sub-task) |
| B (hard) | top of each sub-task iteration | 85% of any axis | Restrict next `generateWithTools` tool list to `[finalize_subtask]` only |
| C | per-tool-call inside sub-task loop | `read_tool_result_artifact` called ≥2× on same `artifactId` | Append guidance nudge to that tool_result |
| D (soft) | after each `dispatch_subtasks` resolves | ≥50% of current batch BUDGET_EXHAUSTED with empty `updatedKdSections`, AND not first batch | Prepend warning to strategist's next dispatch tool_result |
| D (hard) | after each `dispatch_subtasks` resolves | cumulative ≥50% BUDGET_EXHAUSTED OR 2 consecutive batches each ≥80% BUDGET_EXHAUSTED | Restrict strategist tool list to `[complete_analysis, kd_read_toc, kd_read_section]` |
| E (soft) | top of each strategist iteration | 75% of `ticket.totalTokenBudget` | Inject user-text warning into strategist messages (fires once per run) |
| E (hard) | top of each strategist iteration | 95% of `ticket.totalTokenBudget` | Inject E.1 continuation-notes directive AND combine with D's hard-stop tool restriction |
| E.1 | E hard-stop | n/a | Strategist's `complete_analysis.finalAnalysis` must include a `## Continuation Notes` section with 4 subheadings: established / open hypotheses / incomplete threads / suggested next batch — provides resumable structured state for future re-analysis flows |

## Pure functions extracted for testability

- `evaluateSubTaskBudget(usage, config) → 'OK' | 'SOFT_NUDGE' | 'HARD_STOP'` (worst-axis wins)
- `evaluateTicketBudget(totalTokens, config) → 'OK' | 'SOFT_NUDGE' | 'HARD_STOP'`
- `detectArtifactReread(counts, artifactId, threshold) → boolean`
- `evaluateBatchFailureGuard(state, batchResults, config, isFirstBatch) → 'OK' | 'SOFT_NUDGE' | 'HARD_STOP'`

All four live in `services/ticket-analyzer/src/analysis/budget-thresholds.ts` and have 19 isolated unit tests.

## Configuration

New AppSetting `orchestrated-v2-budget-config`. Schema defined in `packages/shared-types/src/analysis.ts` with Zod refinements enforcing `softNudgeRatio < hardStopRatio` for all three tiers.

REST endpoints (per-feature pattern, mirroring `analysis-strategy-version`):
- `GET /api/settings/orchestrated-v2-budget-config`
- `PUT /api/settings/orchestrated-v2-budget-config` (ADMIN-only)

Control panel: new "Orchestrated v2 Budget Limits" card in Settings → Analysis tab. All thresholds editable; Save/Reset disabled when pristine.

MCP tool pair deferred to #475 because exposing `set_*` to the analyzer's caller registry would let the analysis agent re-configure its own cost limits — unacceptable foot-gun.

## Test plan

- [x] Unit tests for `budget-thresholds.ts` (19 tests covering all 4 evaluators + boundary cases)
- [x] Integration tests for `runSubTaskLoop` (re-read detector, soft-nudge, hard-stop)
- [x] Integration tests for `runOrchestratedV2` (batch-failure guard hard-stop, ticket budget hard-stop, continuation-summary directive injection)
- [x] `pnpm typecheck` clean across the entire monorepo
- [x] All affected packages' tests pass (ticket-analyzer 247, copilot-api 125, control-panel 15)
- [ ] Manual replay of ticket #49 (`cf1b96e8`) with tightened cap — confirms clean hard-stop with continuation summary in the analysis. **(Pending: requires hitting dev server with admin session)**
- [ ] Manual control-panel verification — card renders, save persists, reload reflects DB state. **(Pending: requires running dev server)**

## Commits

14 atomic commits, each with TDD discipline (failing test → impl → passing test → commit) where applicable. Two of those are `fix:` commits applied in response to code-quality review feedback during implementation:
- `25453c7` — added missing `strategistGuard` cross-field refine
- `f8882e5` — documented the `>` (not `>=`) boundary deviation in batch guard + added boundary-coverage test

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: services/ticket-analyzer/src/analysis/orchestrated-v2.ts [hunk 1]
```diff
@@ -89,12 +92,17 @@ const STRATEGIST_MAX_TOKENS = 8192;
...
+ * Hard-coded fallback values for the orchestrated-v2 sub-task budget. These are
+ * the schema defaults from `OrchestratedV2BudgetConfigSchema` — kept here for
+ * reference. Live runtime values come from `resolveOrchestratedV2BudgetConfig(db)`
+ * and are passed through `runOrchestratedV2` -> `runSubTaskLoop` via `budgetConfig`.
+ */
+const DEFAULT_SUB_TASK_BUDGET = {
+  iterationCap: 8,
+  tokenBudget: 50_000,
+  callBudget: 20,
+} as const;
```
[the remark this answers is not in the document]
Comment (siir) [reply services/ticket-analyzer/src/analysis/orchestrated-v2.ts]: Good catch — `DEFAULT_SUB_TASK_BUDGET` was never actually referenced; it was a documentation anchor introduced when the original hardcoded constants were removed in T10, but the schema in `@bronco/shared-types` is the real source of truth. Replaced with a one-line comment pointing to the schema and resolver path. Fixed in `ba6150e`.

File: services/ticket-analyzer/src/analysis/orchestrated-v2.ts [hunk 2]
```diff
@@ -1007,6 +1118,75 @@ export async function runOrchestratedV2(
     const orchestrationId = randomUUID();
     appLog.info(`Orchestrated analysis iteration ${i + 1}/${orchMaxIterations}`, { ticketId, iteration: i + 1, orchestrationId }, ticketId, 'ticket');
 
+    // Layer E: ticket-level total-token budget evaluation
+    const totalTokensSoFar = orchTotalInputTokens + orchTotalOutputTokens;
+    const ticketVerdict = evaluateTicketBudget(totalTokensSoFar, budgetConfig.ticket);
+
```
[the remark this answers is not in the document]
Comment (siir) [reply services/ticket-analyzer/src/analysis/orchestrated-v2.ts]: Real correctness gap — at e.g. 80% with a 5-sub-task batch each permitted 50k tokens, the worst case would burn 250k more before the next iteration's check noticed, blowing past the 300k cap by 50%+. Fixed in `63e2910` by adding a pre-batch overshoot guard. After `dispatch_subtasks` resolves, but before the batch executes, the run computes worst-case batch cost as `subtaskCount × subTask.tokenBudget` and aborts if `current + worst-case > totalTokenBudget` (or if top-of-iter is already HARD_STOP). On abort: synthesize a `tool_result` for the dispatch call to keep the conversation well-formed, transition `ticketHardStopActive` (injecting the `## Continuation Notes` directive if it's the first crossing — extracted as a closure so both code paths share it), and `continue` to the next outer iteration where the existing top-of-iter logic forces the restricted tool list. Added an integration test asserting that the sub-task `generateWithTools` path is NEVER invoked when the guard trips.
